dll/cpl/sysdm/general.c

```diff
@@ -417,14 +419,330 @@ static VOID SetProcSpeed(HWND hwnd, HKEY hKey, LPTSTR Value, UINT uID)
     }
 }
 
+static
+BOOL
+IsComma(
+    _In_ WCHAR chr)
+{
+    return (chr <= L' ' || chr == L'.' || chr == L',');
+}
+
+/*
+ * Trim dots and commas
+ */
+static
+VOID
+TrimCommas(
+    _Inout_ PWSTR pStr)
+{
+    SIZE_T Length;
+    UINT i = 0;
+
+    if (!pStr)
+        return;
+
+    Length = wcslen(pStr);
+    if (Length == 0)
+        return;
+
+    /* Trim leading characters */
+    while (i < Length && IsComma(pStr[i]))
+    {
+        i++;
+    }
+
+    if (i > 0)
+    {
+        Length -= i;
+        memmove(pStr, pStr + i, (Length + 1) * sizeof(WCHAR));
+    }
+
+    /* Trim trailing characters */
+    while (Length && IsComma(pStr[Length-1]))
+    {
+        pStr[Length-1] = L'\0';
+        --Length;
+    }
+}
+
+/*
+ * Case insensitive variant of wcsstr
+ */
+static
+PWCHAR wcsistr(PWSTR s1, PCWSTR s2)
+{
+    PWCHAR result = NULL;
+    PWSTR ls1, ls2;
+    SIZE_T l1, l2;
+    UINT i;
+
+    if (!s1 || !s2)
+    {
+        return NULL;
+    }
+
+    l1 = wcslen(s1) + 1;
+    l2 = wcslen(s2) + 1;
+    ls1 = HeapAlloc(GetProcessHeap(), 0, l1 * sizeof(WCHAR));
+    ls2 = HeapAlloc(GetProcessHeap(), 0, l2 * sizeof(WCHAR));
+
+    if (ls1 && ls2)
+    {
+        i = 0;
+        while (*(s1 + i))
+        {
+            *(ls1 + i) = towlower(*(s1 + i));
+            ls1[++i] = L'\0';
+        }
+
+        i = 0;
+        while (*(s2 + i))
+        {
+            *(ls2 + i) = towlower(*(s2 + i));
+            ls2[++i] = L'\0';
+        }
+
+        result = wcsstr(ls1, ls2);
+
+        // workaround for commas
+        if (!result && ls2[l2 - 2] == L' ' && wcschr(ls1, L',') != NULL)
+        {
+            ls2[l2 - 2] = L',';
+            result = wcsstr(ls1, ls2);
+            ls2[l2 - 2] = L' ';
+        }
+
+        if (result)
+        {
+            result = s1 + (result - ls1);
+        }
+    }
+    if (ls1)
+    {
+        HeapFree(GetProcessHeap(), 0, ls1);
+    }
+    if (ls2)
+    {
+        HeapFree(GetProcessHeap(), 0, ls2);
+    }
+    return result;
+}
+
+/*
+ * Replaces full word with another shorter word
+ */
+static
+VOID wcsrep(
```

Comment:
It's a shame this is not a c++ file.

Reply:
Yeah indeed, so I have to implement all that functions.